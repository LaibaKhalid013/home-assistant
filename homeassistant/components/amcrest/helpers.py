"""Helpers for amcrest component."""

from __future__ import annotations

import logging
from typing import TYPE_CHECKING

from .const import DOMAIN

if TYPE_CHECKING:
    from homeassistant.helpers.typing import UndefinedType


def service_signal(service: str, *args: str) -> str:
    """Encode signal."""
    return "_".join([DOMAIN, service, *args])


def log_update_error(
    logger: logging.Logger,
    action: str,
    name: str | UndefinedType | None,
    entity_type: str,
    error: Exception,
    level: int = logging.ERROR,
) -> None:
    """Log an update error."""
    logger.log(
        level,
        "Could not %s %s %s due to error: %s",
        action,
        name,
        entity_type,
        error.__class__.__name__,
    )
