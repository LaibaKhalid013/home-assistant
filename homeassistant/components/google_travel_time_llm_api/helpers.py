"""Helpers for Google Time Travel integration."""

import logging

from googlemaps import Client

from homeassistant.core import HomeAssistant

_LOGGER = logging.getLogger(__name__)


def validate_config_entry(hass: HomeAssistant, api_key: str) -> None:
    """Return whether the config entry data is valid."""
    try:
        Client(api_key, timeout=10)
    except ValueError as value_error:
        _LOGGER.error("Malformed API key")
        raise InvalidApiKeyException from value_error


class InvalidApiKeyException(Exception):
    """Invalid API Key Error."""
