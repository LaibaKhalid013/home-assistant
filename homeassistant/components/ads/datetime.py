"""Support for ADS DateTime."""

from __future__ import annotations

from datetime import UTC, datetime, timedelta

import pyads
import voluptuous as vol

from homeassistant.components.datetime import (
    PLATFORM_SCHEMA as DATETIME_PLATFORM_SCHEMA,
    DateTimeEntity,
)
from homeassistant.const import CONF_NAME
from homeassistant.core import HomeAssistant
import homeassistant.helpers.config_validation as cv
from homeassistant.helpers.entity_platform import AddEntitiesCallback
from homeassistant.helpers.typing import ConfigType, DiscoveryInfoType

from .. import ads
from . import AdsEntity

SCAN_INTERVAL = timedelta(seconds=3)
DEFAULT_NAME = "ADS DateTime"

CONF_ADS_VAR = "adsvar"

PLATFORM_SCHEMA = DATETIME_PLATFORM_SCHEMA.extend(
    {
        vol.Optional(CONF_ADS_VAR): cv.string,
        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
    }
)


def setup_platform(
    hass: HomeAssistant,
    config: ConfigType,
    add_entities: AddEntitiesCallback,
    discovery_info: DiscoveryInfoType | None = None,
) -> None:
    """Set up the DateTime platform for ADS."""

    ads_hub = hass.data.get(ads.DATA_ADS)
    ads_var = config.get(CONF_ADS_VAR)
    name = config[CONF_NAME]

    add_entities(
        [
            AdsDateTime(
                ads_hub,
                ads_var,
                name,
            )
        ]
    )


class AdsDateTime(AdsEntity, DateTimeEntity):
    """Representation of ADS DateTime entity."""

    def __init__(self, ads_hub, ads_var, name):
        """Initialize AdsDateTime entity."""
        super().__init__(ads_hub, name, ads_var)
        self._datetime = None

    async def async_added_to_hass(self) -> None:
        """Register device notification."""

        if self._ads_var is not None:
            await self.async_initialize_device(self._ads_var, pyads.PLCTYPE_DT)
        self.async_schedule_update_ha_state(True)

    @property
    def should_poll(self) -> bool:
        """Return True if the entity should be polled."""
        return True

    async def async_set_value(self, value: datetime) -> None:
        """Set the date/time."""
        new_datetime = value
        await self.hass.async_add_executor_job(self._write_value, new_datetime)
        self._datetime = new_datetime
        self._attr_native_value = new_datetime
        self.async_write_ha_state()

    def _write_value(self, dt_value: datetime) -> None:
        # Convert Python datetime to epoch time (seconds since 1970-01-01)
        epoch_time = int(dt_value.timestamp())
        # Write the epoch time to the PLC as an integer
        self._ads_hub.write_by_name(self._ads_var, epoch_time, pyads.PLCTYPE_DT)

    async def async_update(self) -> None:
        """Retrieve the latest state from the ADS device."""
        epoch_value = self._ads_hub.read_by_name(self._ads_var, pyads.PLCTYPE_DT)
        self._datetime = datetime.fromtimestamp(epoch_value, tz=UTC)
        self._attr_native_value = self._datetime
        self.async_write_ha_state()
