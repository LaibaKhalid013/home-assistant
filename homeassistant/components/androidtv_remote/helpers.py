"""Helper functions for Android TV Remote integration."""

from __future__ import annotations

from typing import TYPE_CHECKING

from androidtvremote2 import AndroidTVRemote

from homeassistant.helpers.storage import STORAGE_DIR

from .const import CONF_ENABLE_IME, CONF_ENABLE_IME_DEFAULT_VALUE

if TYPE_CHECKING:
    from homeassistant.config_entries import ConfigEntry
    from homeassistant.core import HomeAssistant


def create_api(hass: HomeAssistant, host: str, enable_ime: bool) -> AndroidTVRemote:
    """Create an AndroidTVRemote instance."""
    return AndroidTVRemote(
        client_name="Home Assistant",
        certfile=hass.config.path(STORAGE_DIR, "androidtv_remote_cert.pem"),
        keyfile=hass.config.path(STORAGE_DIR, "androidtv_remote_key.pem"),
        host=host,
        loop=hass.loop,
        enable_ime=enable_ime,
    )


def get_enable_ime(entry: ConfigEntry) -> bool:
    """Get value of enable_ime option or its default value."""
    return entry.options.get(CONF_ENABLE_IME, CONF_ENABLE_IME_DEFAULT_VALUE)
