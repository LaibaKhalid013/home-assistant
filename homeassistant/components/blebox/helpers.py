"""Blebox helpers."""

from __future__ import annotations

from typing import TYPE_CHECKING

import aiohttp

from homeassistant.helpers.aiohttp_client import (
    async_create_clientsession,
    async_get_clientsession,
)

if TYPE_CHECKING:
    from homeassistant.core import HomeAssistant


def get_maybe_authenticated_session(
    hass: HomeAssistant, password: str | None, username: str | None
) -> aiohttp.ClientSession:
    """Return proper session object."""
    if username and password:
        auth = aiohttp.BasicAuth(login=username, password=password)
        return async_create_clientsession(hass, auth=auth)

    return async_get_clientsession(hass)
