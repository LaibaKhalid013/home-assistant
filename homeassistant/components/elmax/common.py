"""Elmax integration common classes and utilities."""

from __future__ import annotations

import ssl

from elmax_api.model.endpoint import DeviceEndpoint
from elmax_api.model.panel import PanelEntry
from packaging import version

from homeassistant.core import callback
from homeassistant.helpers.device_registry import DeviceInfo
from homeassistant.helpers.dispatcher import async_dispatcher_connect
from homeassistant.helpers.update_coordinator import CoordinatorEntity

from .const import (
    DOMAIN,
    ELMAX_LOCAL_API_PATH,
    MIN_APIV2_SUPPORTED_VERSION,
    SIGNAL_PANEL_UPDATE,
)
from .coordinator import ElmaxCoordinator


def get_direct_api_url(host: str, port: int, use_ssl: bool) -> str:
    """Return the direct API url given the base URI."""
    schema = "https" if use_ssl else "http"
    return f"{schema}://{host}:{port}/{ELMAX_LOCAL_API_PATH}"


def build_direct_ssl_context(cadata: str) -> ssl.SSLContext:
    """Create a custom SSL context for direct-api verification."""
    context = ssl.SSLContext(protocol=ssl.PROTOCOL_TLS_CLIENT)
    context.check_hostname = False
    context.verify_mode = ssl.CERT_REQUIRED
    context.load_verify_locations(cadata=cadata)
    return context


def check_local_version_supported(api_version: str | None) -> bool:
    """Check whether the given API version is supported."""
    if api_version is None:
        return False
    return version.parse(api_version) >= version.parse(MIN_APIV2_SUPPORTED_VERSION)


class DirectPanel(PanelEntry):
    """Helper class for wrapping a directly accessed Elmax Panel."""

    def __init__(self, panel_uri):
        """Construct the object."""
        super().__init__(panel_uri, True, {})

    def get_name_by_user(self, username: str) -> str:
        """Return the panel name."""
        return f"Direct Panel {self.hash}"


class ElmaxEntity(CoordinatorEntity[ElmaxCoordinator]):
    """Wrapper for Elmax entities."""

    _last_state: DeviceEndpoint = None

    def __init__(
        self,
        elmax_device: DeviceEndpoint,
        panel_version: str,
        coordinator: ElmaxCoordinator,
    ) -> None:
        """Construct the object."""
        super().__init__(coordinator=coordinator)
        self._device = elmax_device
        self._attr_unique_id = elmax_device.endpoint_id
        self._attr_name = elmax_device.name
        self._attr_device_info = DeviceInfo(
            identifiers={(DOMAIN, coordinator.panel_entry.hash)},
            name=coordinator.panel_entry.get_name_by_user(
                coordinator.http_client.get_authenticated_username()
            ),
            manufacturer="Elmax",
            model=panel_version,
            sw_version=panel_version,
        )

    async def async_added_to_hass(self) -> None:
        """Register push notifications callbacks if available."""
        self.async_on_remove(
            async_dispatcher_connect(
                self.hass,
                f"{SIGNAL_PANEL_UPDATE}-{self.coordinator.panel_entry.hash}-{self._device.endpoint_id}",
                self._handle_update,
            )
        )
        return await super().async_added_to_hass()

    @callback
    def _handle_update(self, endpoint_status: DeviceEndpoint) -> None:
        self._last_state = endpoint_status
        self.async_write_ha_state()

    @property
    def available(self) -> bool:
        """Return if entity is available."""
        return super().available and self.coordinator.panel_entry.online
