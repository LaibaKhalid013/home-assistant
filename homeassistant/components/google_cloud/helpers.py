"""Helper classes for Google Cloud integration."""

from __future__ import annotations

import logging
from types import MappingProxyType
from typing import Any

from google.cloud import texttospeech
import voluptuous as vol

from homeassistant.components.tts import CONF_LANG
import homeassistant.helpers.config_validation as cv
from homeassistant.helpers.selector import (
    NumberSelector,
    NumberSelectorConfig,
    SelectSelector,
    SelectSelectorConfig,
    SelectSelectorMode,
)

from .const import (
    CONF_ENCODING,
    CONF_GAIN,
    CONF_GENDER,
    CONF_KEY_FILE,
    CONF_PITCH,
    CONF_PROFILES,
    CONF_SPEED,
    CONF_TEXT_TYPE,
    CONF_VOICE,
    DEFAULT_LANG,
)

_LOGGER = logging.getLogger(__name__)

DEFAULT_GENDER = texttospeech.SsmlVoiceGender.NEUTRAL.name
DEFAULT_VOICE = ""
DEFAULT_ENCODING = texttospeech.AudioEncoding.MP3.name
DEFAULT_SPEED = 1.0
DEFAULT_PITCH = 0
DEFAULT_GAIN = 0
DEFAULT_PROFILES: list[str] = []
DEFAULT_TEXT_TYPE = "text"

# https://cloud.google.com/text-to-speech/docs/audio-profiles
SUPPORTED_PROFILES = [
    "wearable-class-device",
    "handset-class-device",
    "headphone-class-device",
    "small-bluetooth-speaker-class-device",
    "medium-bluetooth-speaker-class-device",
    "large-home-entertainment-class-device",
    "large-automotive-class-device",
    "telephony-class-application",
]
SUPPORTED_TEXT_TYPES = ["text", "ssml"]


async def async_tts_voices(
    client: texttospeech.TextToSpeechAsyncClient,
) -> dict[str, list[str]]:
    """Get TTS voice models keyed by language."""
    voices: dict[str, list[str]] = {}
    list_voices_response = await client.list_voices()
    for voice in list_voices_response.voices:
        language_code = voice.language_codes[0]
        if language_code not in voices:
            voices[language_code] = []
        voices[language_code].append(voice.name)
    return voices


def tts_options_schema(
    config_options: MappingProxyType[str, Any], voices: dict[str, list[str]]
):
    """Return schema for TTS options with default values from config or constants."""
    return vol.Schema(
        {
            vol.Optional(
                CONF_GENDER,
                description={"suggested_value": config_options.get(CONF_GENDER)},
                default=DEFAULT_GENDER,
            ): SelectSelector(
                SelectSelectorConfig(
                    mode=SelectSelectorMode.DROPDOWN,
                    options=list(texttospeech.SsmlVoiceGender.__members__),
                )
            ),
            vol.Optional(
                CONF_VOICE,
                description={"suggested_value": config_options.get(CONF_VOICE)},
                default=DEFAULT_VOICE,
            ): SelectSelector(
                SelectSelectorConfig(
                    mode=SelectSelectorMode.DROPDOWN,
                    options=["", *sum(voices.values(), [])],
                )
            ),
            vol.Optional(
                CONF_ENCODING,
                description={"suggested_value": config_options.get(CONF_ENCODING)},
                default=DEFAULT_ENCODING,
            ): SelectSelector(
                SelectSelectorConfig(
                    mode=SelectSelectorMode.DROPDOWN,
                    options=list(texttospeech.AudioEncoding.__members__),
                )
            ),
            vol.Optional(
                CONF_SPEED,
                description={"suggested_value": config_options.get(CONF_SPEED)},
                default=DEFAULT_SPEED,
            ): NumberSelector(NumberSelectorConfig(min=0.25, max=4, step=0.05)),
            vol.Optional(
                CONF_PITCH,
                description={"suggested_value": config_options.get(CONF_PITCH)},
                default=DEFAULT_PITCH,
            ): NumberSelector(NumberSelectorConfig(min=-20, max=20, step=0.1)),
            vol.Optional(
                CONF_GAIN,
                description={"suggested_value": config_options.get(CONF_GAIN)},
                default=DEFAULT_GAIN,
            ): NumberSelector(NumberSelectorConfig(min=-96, max=16, step=1)),
            vol.Optional(
                CONF_PROFILES,
                description={"suggested_value": config_options.get(CONF_PROFILES)},
                default=DEFAULT_PROFILES,
            ): SelectSelector(
                SelectSelectorConfig(
                    mode=SelectSelectorMode.DROPDOWN,
                    options=SUPPORTED_PROFILES,
                    multiple=True,
                    sort=False,
                )
            ),
            vol.Optional(
                CONF_TEXT_TYPE,
                description={"suggested_value": config_options.get(CONF_TEXT_TYPE)},
                default=DEFAULT_TEXT_TYPE,
            ): SelectSelector(
                SelectSelectorConfig(
                    mode=SelectSelectorMode.DROPDOWN,
                    options=SUPPORTED_TEXT_TYPES,
                )
            ),
        }
    )


def tts_platform_schema():
    """Return schema for TTS platform."""
    return vol.Schema(
        {
            vol.Optional(CONF_KEY_FILE): cv.string,
            vol.Optional(CONF_LANG, default=DEFAULT_LANG): cv.matches_regex(
                r"[a-z]{2,3}-[A-Z]{2}"
            ),
            **tts_options_schema(config_options={}, voices={}).schema,
            vol.Optional(CONF_VOICE, default=DEFAULT_VOICE): cv.matches_regex(
                r"[a-z]{2,3}-[A-Z]{2}-.*-[A-Z]|"
            ),
        }
    )
