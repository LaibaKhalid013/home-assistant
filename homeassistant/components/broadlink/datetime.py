"""Support for Broadlink sensors."""

from __future__ import annotations

from datetime import datetime, timedelta

from homeassistant.components.datetime import DateTimeEntity
from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant
from homeassistant.helpers.entity_platform import AddEntitiesCallback
from homeassistant.util import dt as dt_util

from .const import DOMAIN
from .entity import BroadlinkEntity


async def async_setup_entry(
    hass: HomeAssistant,
    config_entry: ConfigEntry,
    async_add_entities: AddEntitiesCallback,
) -> None:
    """Set up the Broadlink datetime."""
    device = hass.data[DOMAIN].devices[config_entry.entry_id]
    async_add_entities([BroadlinkDateTime(device)])


class BroadlinkDateTime(BroadlinkEntity, DateTimeEntity):
    """Representation of a Broadlink date and time."""

    _attr_has_entity_name = True
    _attr_native_value: datetime | None = None

    def __init__(self, device) -> None:
        """Initialize the sensor."""
        super().__init__(device)

        self._attr_unique_id = f"{device.unique_id}-datetime"

    def _update_state(self, data) -> None:
        """Update the state of the entity."""
        if data is None or "dayofweek" not in data:
            self._attr_native_value = None
        else:
            now = dt_util.now()
            device_weekday = data["dayofweek"] - 1
            this_weekday = now.weekday()

            if device_weekday != this_weekday:
                days_diff = this_weekday - device_weekday
                if days_diff < 0:
                    days_diff += 7
                now -= timedelta(days=days_diff)

            self._attr_native_value = now.replace(
                hour=data["hour"],
                minute=data["min"],
                second=data["sec"],
            )

    async def async_set_value(self, value: datetime) -> None:
        """Change the value."""
        value = dt_util.as_local(value)
        await self._device.async_request(
            self._device.api.set_time,
            hour=value.hour,
            minute=value.minute,
            second=value.second,
            day=value.weekday() + 1,
        )
        self._attr_native_value = value
        self.async_write_ha_state()
