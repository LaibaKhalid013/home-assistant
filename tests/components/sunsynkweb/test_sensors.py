"""sensor tests."""

from __future__ import annotations

from datetime import timedelta
from typing import Literal
from unittest.mock import AsyncMock

from syrupy.assertion import SnapshotAssertion

from homeassistant.core import HomeAssistant
from homeassistant.helpers.entity_registry import EntityRegistry

from tests.common import MockConfigEntry, async_fire_time_changed


async def next_tick(hass, freezer):
    """Make hass trigger the next update."""
    freezer.tick(timedelta(seconds=31))
    async_fire_time_changed(hass)  # noqa: F821
    await hass.async_block_till_done(wait_background_tasks=True)


async def test_simple_update(
    hass: HomeAssistant,
    entity_registry: EntityRegistry,
    init_integration: MockConfigEntry,
    freezer,
    sensor_keys: tuple[
        Literal["battery_power"],
        Literal["load_power"],
        Literal["grid_power"],
        Literal["pv_power"],
        Literal["state_of_charge"],
        Literal["acc_pv"],
        Literal["acc_load"],
        Literal["acc_grid_import"],
        Literal["acc_grid_export"],
        Literal["acc_battery_charge"],
        Literal["acc_battery_discharge"],
    ],
    snapshot: SnapshotAssertion,
) -> None:
    """Validate that all sensors are registered and that updates propagate."""
    for key in sensor_keys:
        state = hass.states.get(f"sensor.sunsynk_{key}")
        assert state.state == "unknown"

    freezer.tick(timedelta(seconds=10))
    async_fire_time_changed(hass)  # noqa: F821
    await hass.async_block_till_done(wait_background_tasks=True)
    for key in sensor_keys:
        state = hass.states.get(f"sensor.sunsynk_{key}")
        assert state.state == snapshot

    await next_tick(hass, freezer)
    for key in sensor_keys:
        state = hass.states.get(f"sensor.sunsynk_{key}")
        assert state.state == snapshot


async def test_bad_api_return_and_recovery(
    hass: HomeAssistant,
    entity_registry: EntityRegistry,
    init_integration: MockConfigEntry,
    freezer,
    sensor_keys: tuple[
        Literal["battery_power"],
        Literal["load_power"],
        Literal["grid_power"],
        Literal["pv_power"],
        Literal["state_of_charge"],
        Literal["acc_pv"],
        Literal["acc_load"],
        Literal["acc_grid_import"],
        Literal["acc_grid_export"],
        Literal["acc_battery_charge"],
        Literal["acc_battery_discharge"],
    ],
    snapshot: SnapshotAssertion,
) -> None:
    """Verify that if api calls return invalid responses, all sensors are appropriately flagged as unavailable, then recover on normal updates."""
    coordinator = init_integration.runtime_data

    async def raisekeyerror():
        raise KeyError

    coordinator.cache.plants[0].update = raisekeyerror
    await next_tick(hass, freezer)
    for key in sensor_keys:
        state = hass.states.get(f"sensor.sunsynk_{key}")
        assert state.state == snapshot
    coordinator.cache.plants[0].update = AsyncMock()
    await next_tick(hass, freezer)
    for key in sensor_keys:
        state = hass.states.get(f"sensor.sunsynk_{key}")
        assert state.state == snapshot
