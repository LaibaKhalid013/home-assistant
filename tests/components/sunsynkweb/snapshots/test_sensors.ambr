# serializer version: 1
# name: test_bad_api_return_and_recovery
  'unavailable'
# ---
# name: test_bad_api_return_and_recovery.1
  'unavailable'
# ---
# name: test_bad_api_return_and_recovery.10
  'unavailable'
# ---
# name: test_bad_api_return_and_recovery.11
  '0'
# ---
# name: test_bad_api_return_and_recovery.12
  '0'
# ---
# name: test_bad_api_return_and_recovery.13
  '0'
# ---
# name: test_bad_api_return_and_recovery.14
  '0'
# ---
# name: test_bad_api_return_and_recovery.15
  '2'
# ---
# name: test_bad_api_return_and_recovery.16
  '0'
# ---
# name: test_bad_api_return_and_recovery.17
  '0'
# ---
# name: test_bad_api_return_and_recovery.18
  '0'
# ---
# name: test_bad_api_return_and_recovery.19
  '0'
# ---
# name: test_bad_api_return_and_recovery.2
  'unavailable'
# ---
# name: test_bad_api_return_and_recovery.20
  '0'
# ---
# name: test_bad_api_return_and_recovery.21
  '0'
# ---
# name: test_bad_api_return_and_recovery.3
  'unavailable'
# ---
# name: test_bad_api_return_and_recovery.4
  'unavailable'
# ---
# name: test_bad_api_return_and_recovery.5
  'unavailable'
# ---
# name: test_bad_api_return_and_recovery.6
  'unavailable'
# ---
# name: test_bad_api_return_and_recovery.7
  'unavailable'
# ---
# name: test_bad_api_return_and_recovery.8
  'unavailable'
# ---
# name: test_bad_api_return_and_recovery.9
  'unavailable'
# ---
# name: test_simple_update
  '0'
# ---
# name: test_simple_update.1
  '0'
# ---
# name: test_simple_update.10
  '0'
# ---
# name: test_simple_update.11
  '0'
# ---
# name: test_simple_update.12
  '0'
# ---
# name: test_simple_update.13
  '0'
# ---
# name: test_simple_update.14
  '0'
# ---
# name: test_simple_update.15
  '4'
# ---
# name: test_simple_update.16
  '0'
# ---
# name: test_simple_update.17
  '0'
# ---
# name: test_simple_update.18
  '0'
# ---
# name: test_simple_update.19
  '0'
# ---
# name: test_simple_update.2
  '0'
# ---
# name: test_simple_update.20
  '0'
# ---
# name: test_simple_update.21
  '0'
# ---
# name: test_simple_update.3
  '0'
# ---
# name: test_simple_update.4
  '3'
# ---
# name: test_simple_update.5
  '0'
# ---
# name: test_simple_update.6
  '0'
# ---
# name: test_simple_update.7
  '0'
# ---
# name: test_simple_update.8
  '0'
# ---
# name: test_simple_update.9
  '0'
# ---
